Add event processing queue

crazytonyli opened the pull request:
## Issue

This is an attempt to fix some crashes. Here is a part of one crash report:
```
Thread 0 Crashed:
0   libswiftCore.dylib                   0x0000000192a6a198 swift_isUniquelyReferenced_nonNull_native + 0
1   ParselyAnalytics                     0x00000001072a1b1c ParselyAnalytics.EngagedTime.endInteraction() -> () (<compiler-generated>:0)
2   ParselyAnalytics                     0x00000001072a6444 ParselyAnalytics.Parsely.stopEngagement() -> () (Track.swift:63)
```

`EngagedTime.endInteraction` and `Sampler.sendHeartbeats` are two common crash sites.

## Potential Cause

I suspect this is a concurrency issue. There is almost no guarantee of where some of the functions are executed, but most of the public APIs eventually modify the internal state like [`accumulators`](https://github.com/Parsely/AnalyticsSDK-iOS/blob/6f0ab679f0de687e09814d75acac7aee75e48e96/ParselyTracker/Sampler.swift#L34) or [`eventQueue`](https://github.com/Parsely/AnalyticsSDK-iOS/blob/6f0ab679f0de687e09814d75acac7aee75e48e96/ParselyTracker/ParselyTracker.swift#L14).

## Fix

This PR adds a serial queue as the place to modify internal state. This PR changes three kinds of code to make they run inside this serial queue:
1. All "tracking" related public APIs.
2. App lifecycle notification handlers.
3. Timer handlers.

In fact, I reproduce similar crashes in unit tests by accident. You can see the crash by checking out 3a9773b48e2989c1d7754cce679c7b55a3d95b0b and running `ParselyTrackerTests` for a few times. The crash would show up (or not) randomly in different test methods.

Originally those unit tests executed assertions on the test thread. Since these assertions inspected the tracker's internal state, there were concurrency issues here: the state was modified from the newly added serial queue, but at the same time it was read from the unit test thread. This concurrency issue is fixed by moving the unit test's assertion into the tracker serial queue, and the unit test passes reliably now.

BTW, it might be easier to review this PR commit by commit.

crazytonyli committed:
Add event processing queue to avoid potential race conditions

crazytonyli committed:
Reimplement flush timer in Combine

So that we can take advantage of the DispatchQueue support in Combine

crazytonyli committed:
Resume and suspend in the event processor queue

crazytonyli committed:
Do not use an global queue

The function is now executed in the event processor queue.

crazytonyli committed:
Use the event processor queue in `Sampler`

crazytonyli committed:
Fix concurrenty issues in some unit tests

mokagio commented on ParselyDemo.xcodeproj/project.pbxproj:
```diff
@@ -366,6 +366,7 @@
 				F433339220F0087E004BB6F5 /* Sources */,
 				F433339320F0087E004BB6F5 /* Frameworks */,
 				F433339420F0087E004BB6F5 /* Resources */,
+				50A891132731B54A9972347D /* [CP] Embed Pods Frameworks */,
```
LOL. I just opened #7 to remove CocoaPods 😳 .

I'll rebase it on top of your work and bring in Nimble via SPM there.

mokagio commented on ParselyTracker/ParselyTracker.swift:
```diff
@@ -66,12 +72,25 @@ public class Parsely {
+    ) {
         var _siteId = siteId
         if (_siteId == "") {
             _siteId = self.apikey
```
```suggestion
            _siteId = apikey
```

mokagio commented on ParselyTracker/ParselyTracker.swift:
```diff
@@ -66,12 +72,25 @@ public class Parsely {
+        extraData: Dictionary<String, Any>? = nil,
+        siteId: String = ""
+    ) {
+        eventProcessor.async {
+            self._trackPageView(url: url, urlref: urlref, metadata: metadata, extraData: extraData, siteId: siteId)
+        }
+    }
+
+    private func _trackPageView(
```
Have you considered keeping the code inside `eventProcessor.async { }`?

Nitpick. This approach makes the body of the private version clean but... I don't know... I feel like all these parameters redefined just for one call make the code heavier 🤷‍♂️ 😅 

Or is there a runtime/performance reason for this?

No worries either way.

crazytonyli replied:
I wanted to make this PR easier to review. So, I avoided making too many changes to the existing code. I can follow up you suggestions in a separate PR though. What do you think?

mokagio replied:
Sure. No worries. Super low priority. I'm just glad it was a style thing and that it wasn't a code behavior that I didn't know about. 

If there's time to do that, good, otherwise, no dramas. You have more important things to tackle that this nitpick 👍

mokagio commented on ParselyTrackerTests/ParselyTrackerTests.swift:
```diff
@@ -20,12 +22,15 @@ class ParselyTrackerTests: ParselyTestCase {
         XCTAssertEqual(parselyTestTracker.eventQueue.length(), 0,
                        "eventQueue should be empty immediately after initialization")
         parselyTestTracker.trackPageView(url: testUrl, urlref: testUrl, metadata: nil, extraData: nil)
+        expectParselyState(self.parselyTestTracker.eventQueue.length()).toEventually(equal(1))
         XCTAssertEqual(parselyTestTracker.eventQueue.length(), 1,
                        "A call to Parsely.trackPageView should add an event to eventQueue")
```
The XCTest assertion is now redundant.

```suggestion
        expect(self.parselyTestTracker.eventQueue.length())
            .toEventually(
                equal(1),
                description: "A call to Parsely.trackPageView should add an event to eventQueue"
            )
```

mokagio commented on ParselyTracker/ParselyTracker.swift:
```diff
@@ -154,7 +204,18 @@ public class Parsely {
+        }
+    }
+
+    /// After given `seconds`, invoke the given target-action in the event processor queue.
+    func scheduleEventProcessing(inSeconds seconds: Double, target: AnyObject, selector: Selector) -> Cancellable {
+        Just(0)
+            .delay(for: .seconds(seconds), scheduler: eventProcessor)
+            .sink { _ in
+                _ = target.perform(selector)
+            }
     }
```
I'm a bit unsure about having other objects using this method to schedule their own `Selector`s on the tracker's `DispatchQueue`.

I came to this only when I read the diff in the other objects, though. I don't know if there's something we can do to make the naming clearer or if it's a responsibility thing 🤔 

Have you considered extracting this logic in an object (or a `DispatchQueue` extension) and pass _that_ around, to make the different responsibilities clearer?

Anyways, this is more of a design _question_. Not a blocker for this PR.

crazytonyli replied:
The purpose of this function is to make it more explicit that this is _the_ function for scheduling event processing action. It's hard to convey the same idea if there is a "schedule event processing" function on all `DispatchQueue` instances, I think?

Another reason I added this function in this type is because it's referenced pretty much everywhere in the SDK. I thought may as well use it as the "event processing scheduler".

mokagio replied:
> The purpose of this function is to make it more explicit that this is the function for scheduling event processing action.

👌 

> Another reason I added this function in this type is because it's referenced pretty much everywhere in the SDK. I thought may as well use it as the "event processing scheduler".

Yeah. That's fair.

>  It's hard to convey the same idea if there is a "schedule event processing" function on all `DispatchQueue` instances, I think?

Good point on the `DispatchQueue` extension, because then any `DispatchQueue` instance could be used to queue events and that's not what we want. In the context of what you said above regarding the current design of the SDK, I can see how having _another_ object to pass around (or access via _another_ shared instance 😱) would make things more complex... unless the rest of the code was tidied up, too, which is definitely out of scope right now.

mokagio reviewed (commented):
Looking good.

I'm not approving yet because I'd like to have CI run some stress tests. I hope to use the commit you highlighted as the benchmark where the tests fail and show how these changes fix them.

mokagio commented:
>  I hope to use the commit you highlighted as the benchmark where the tests fail and show how these changes fix them.

By the way, on my machine when running the tests via `xcodebuild` with a test plan that repeats them 10 time, I get failures on the `testBackoff`:

```
SamplerTests
    ✖ testBackoff, XCTAssertEqual failed: ("13.68") is not equal to ("13.65") - Heartbeat interval should increase by the expected amount after a single heartbeat
    ✖ testBackoff, XCTAssertEqual failed: ("13.68") is not equal to ("13.65") - Heartbeat interval should increase by the expected amount after a single heartbeat
    ✖ testBackoff, XCTAssertEqual failed: ("13.67") is not equal to ("13.65") - Heartbeat interval should increase by the expected amount after a single heartbeat
    ✖ testBackoff, XCTAssertEqual failed: ("13.68") is not equal to ("13.65") - Heartbeat interval should increase by the expected amount after a single heartbeat
    ✖ testBackoff, XCTAssertEqual failed: ("13.67") is not equal to ("13.65") - Heartbeat interval should increase by the expected amount after a single heartbeat
    ✖ testBackoff, XCTAssertEqual failed: ("13.67") is not equal to ("13.65") - Heartbeat interval should increase by the expected amount after a single heartbeat
    ✖ testBackoff, XCTAssertEqual failed: ("13.67") is not equal to ("13.65") - Heartbeat interval should increase by the expected amount after a single heartbeat
    ✔ testBackoff (12.502 seconds)
    ✖ testBackoff, XCTAssertEqual failed: ("13.67") is not equal to ("13.65") - Heartbeat interval should increase by the expected amount after a single heartbeat
    ✖ testBackoff, XCTAssertEqual failed: ("13.68") is not equal to ("13.65") - Heartbeat interval should increase by the expected amount after a single heartbeat
```

mokagio commented:
The PR I put up to "stress test" the tip of this branch, #8, had 3 failures on the same `SamplerTests.testBackoff()` test that I experienced.

I haven't looked into it yet to verify whether it's related to the changes here.

mokagio commented:
The PR I created to "stress test" `3a9773b`, https://github.com/Parsely/AnalyticsSDK-iOS/pull/75 (accidentally on the Parsely repo), had [many test failures](https://github.com/Parsely/AnalyticsSDK-iOS/actions/runs/4699442389/jobs/8332904928?pr=75#step:6:1164) _as expected_. 👍  

![image](https://user-images.githubusercontent.com/1218433/232038963-5136e93d-d8b3-4a0c-92b5-99b49bea9c11.png)
(_790 test are the 79 test in the suite each run 10 times_)

![image](https://user-images.githubusercontent.com/1218433/232038973-4bfdde23-539a-4b3e-b877-d55eb4598e75.png)

That includes failures in `ParselyTrackerTests` which is were the new `DispatchQueue` is used.

mokagio commented:
I run another stress test on the base branch. [They all passed](https://github.com/Automattic/AnalyticsSDK-iOS/actions/runs/4699619615/jobs/8333309852?pr=10). 

That is should be consistent with the changes done here. This changes introduce the `DispatchQueue` and using it properly is what may address our race conditions. Before these changes, the code use `Timer.scheduledTimer` which I guess is more stable for how the tests run one after the other 🤔

mokagio commented:
> I haven't looked into it yet to verify whether it's related to the changes here.

`SamplerTests.testBackoff` is definitely a test that touches the code changed in this PR.

crazytonyli commented:
I've seen the same `testBackOff` failure on [this PR](https://github.com/Parsely/AnalyticsSDK-iOS/actions/runs/4694729854/jobs/8323157918?pr=72#step:6:156) which didn't change any runtime behaviour. So, I suspect this test was flaky, but the change in this PR probably made it worse.

crazytonyli committed:
Remove redundant test assertions

crazytonyli committed:
Make the notification handling `OperationQueue` serial queue

I don't think this new line is necessarily required, since the
underlying queue is a serial queue. But I guess it doesn't hurt to
explicitily mark it as a serial queue.

crazytonyli committed:
Handle API responses within the event processor queue

mokagio commented on ParselyTrackerTests/ParselyTrackerTests.swift:
```diff
@@ -22,72 +22,57 @@ class ParselyTrackerTests: ParselyTestCase {
-        XCTAssert(testVideo.isPlaying,
-                  "After a call to Parsely.trackPlay, the tracked video should have its isPlaying flag set")
+        XCTAssertEqual(parselyTestTracker.eventQueue.length(), 1, "A call to Parsely.trackPlay should add an event to eventQueue")
+        try XCTAssert(
```
Nitpick: I find `XCTAssertTrue` clearer than `XCTAssert`, even though they do the same thing.

<img width="482" alt="Screenshot 2023-04-17 at 2 06 48 pm" src="https://user-images.githubusercontent.com/1218433/232375612-bd2aa4cb-0b34-496a-99c1-e9bbebaab649.png">

```suggestion
        try XCTAssertTrue(
```

mokagio commented on ParselyTrackerTests/ParselyTrackerTests.swift:
```diff
@@ -22,72 +22,57 @@ class ParselyTrackerTests: ParselyTestCase {
-        let trackedVideos: Dictionary<String, TrackedVideo> = videoManager.trackedVideos
-        XCTAssertEqual(trackedVideos.count, 0,
-                       "A call to Parsely.resetVideo should remove an tracked video from the video manager")
+        // A call to Parsely.resetVideo should remove an tracked video from the video manager
```
```suggestion
        // A call to Parsely.resetVideo should remove a tracked video from the video manager
```

mokagio reviewed (approved):
> I've seen the same testBackOff failure on [this PR](https://github.com/Parsely/AnalyticsSDK-iOS/actions/runs/4694729854/jobs/8323157918?pr=72#step:6:156) which didn't change any runtime behaviour. So, I suspect this test was flaky, but the change in this PR probably made it worse.

As I read that, I though "maybe the switch to using Combine adds some intrinsic uncertainty on the exact duration of the delay. I initially went for relaxing the assertion by adding a range of accuracy (#11) then it occurred to me to check [the docs for `.delay(for...`](https://developer.apple.com/documentation/combine/just/delay(for:tolerance:scheduler:options:)) and learned a few interesting things:

- The Apple docs state that the delay interval has a range of accuracy. In their example, `.seconds(3)` amounts to "3 seconds (±0.5 seconds)"
- The method offers a `tolerance` parameter: "The allowed tolerance in delivering delayed events. The Delay publisher may deliver elements this much sooner or later than the interval specifies."

I tried `.milliseconds(1)` and still got failures, including a 13.61 interval (0.04 delta) which is the biggest one I ever saw:

<img width="489" alt="image" src="https://user-images.githubusercontent.com/1218433/232390100-1ebf8406-62ab-496a-bc4a-e7202fb201ca.png">

I then tried `.nanoseconds(1)` and still got flakiness:

<img width="1266" alt="image" src="https://user-images.githubusercontent.com/1218433/232390977-860447ed-df53-41a6-bac0-0b46a8abba4a.png">

At that point, I gave up and decided that adding an accuracy range to the test was the only approach. #11's description goes more into more details.

crazytonyli committed:
Replace `XCTAssert` with `XCTAssertTrue`

aimeevan commented:
Hi All - One of the customers affected by this has asked for daily updates on our progress (even if it means sharing that there are no updates). 

What is appropriate for me to share at the moment? My last update to them was 'our developers have begun to initiate the process of analyzing the SDK issue.' Is there something new I can add to this?

mokagio commented:
Hi @aimeevan 👋  I think that sums that up, wish we had more to share at this point but without an app in our hands were we can replicate the issue and see if our changes fix it, it's hard to be specific.
